File: code/python/mandel-gen-input.py
import numpy as np
import os

xmin, xmax = np.float64(-2.5), np.float64(1.0)
ymin, ymax = np.float64(-1.25), np.float64(1.25)
nx = int(os.environ.get('INPUTMAT_ROWS'))
ny = int(os.environ.get('INPUTMAT_COLS'))
maxiter = 100


def linspace(start, end, n):
    lspace = [start]
    distance = (end - start) / np.float64(n - 1)
    next = start
    while np.abs(next - end) > 0.000000000001:
        next += distance
        lspace.append(next)
    return lspace


X = linspace(xmin, xmax, nx)
Y = linspace(ymin, ymax, ny)

def compute_all_y(x):
    return [complex(x, y) for y in Y]

Z = list(map(compute_all_y, X))

File: code/go/mandel-gen-input.go
package main

import "math"
import "math/cmplx"
import "os"
import "strconv"
import "sync"
import "runtime"

var rows, err1 = strconv.Atoi(os.Getenv("INPUTMAT_ROWS"))
var columns, err2 = strconv.Atoi(os.Getenv("INPUTMAT_COLS"))

func mandelbrotorbit(c complex128) int {
	z := complex(0, 0) + c
	for i := 1; i <= 100; i++ {
		if cmplx.Abs(z) > 2 {
			return i - 1
		}
		z = cmplx.Pow(z, 2) + c
	}
	return 100
}

func mandelbrot(inputmat [][]complex128) [][]int {
	runtime.GOMAXPROCS(8)
	var wg sync.WaitGroup
	wg.Add(rows)

	outputmat := make([][]int, rows)
	for i := range outputmat {
		outputmat[i] = make([]int, columns)
	}

	for i := 0; i < rows; i++ {
		go func(i int) {
			for j := 0; j < columns; j++ {
				outputmat[i][j] = mandelbrotorbit(inputmat[i][j])
			}
			wg.Done()
		}(i)
	}
	wg.Wait()
	return outputmat
}

func linspace(start float64, end float64, n int) []float64 {
	var linspace []float64
	linspace = append(linspace, start)
	distance := (end - start) / float64(n-1)
	next := start
	for math.Abs(next-end) > 0.000000000001 {
		next = next + distance
		linspace = append(linspace, next)
	}
	return linspace
}

func clinspace(start complex128, end complex128, m int, n int) [][]complex128 {
	realParts := linspace(real(start), real(end), m)
	imagLinspace := linspace(imag(start), imag(end), n)
	var cmplxParts []float64
	for im := range imagLinspace {
		cmplxParts = append(cmplxParts, imagLinspace[im])
	}

	var cmplxLinspace [][]complex128
	for r := range realParts {
		var cmplxRow []complex128
		for c := range cmplxParts {
			cmplxRow = append(cmplxRow, complex(realParts[r], cmplxParts[c]))
		}
		cmplxLinspace = append(cmplxLinspace, cmplxRow)
	}
	return cmplxLinspace
}

func main() {
	clinspace(-2.5-1.25i, 1.0+1.25i, rows, columns)
}
